proposals/NNNN-resource-arrays.md

```diff
@@ -0,0 +1,356 @@
+
+For example, the handle initialization for the resource that is referenced by `P[3]`
+will have lower bound `0`, upper bound `49`, and index `38 = 0 + 7*5 + 3`:
+
+```
+call %dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 0, i32 49, i32 0, i8 1 }, i32 38, i1 false)
+```
+https://godbolt.org/z/YejsdsTKc
+
+At the same time, `P` is a local variable and as such it is editable. It should
+be initialized by copy-in array semantics to refer to resource handles from the
+original larger array `N`, but its individual array elements can be overridden,
+and that must not affect the global array `N` or the resources it contains in
+any way.
+
+#### Example where DXC Does Not Handle Local Arrays Correctly
+
+Handling of local arrays in DXC is notoriously buggy. For example, in the
+following case the shader should write `1` to `Y` and `2` to `X`, but instead
+both writes go to `Y`, so the end result is that `X` is unused and is optimized
+away.
+
+```
+RWBuffer<int> X : register(u0);
+RWBuffer<int> Y : register(u1);
+
+void SomeFn(RWBuffer<int> Arr[2], uint Idx, int Val0) {
+  Arr[0] = Y;
+  Arr[0][Idx] = Val0;
+}
+
+[numthreads(4,1,1)]
+void main(uint GI : SV_GroupIndex) {
+  RWBuffer<int> Arr[2] = {X, Y};
+  SomeFn(Arr, GI, 1);
+  Arr[0][GI] = 2;
+}
+```
+https://godbolt.org/z/YxM5M6zqo
+
+A brief inspection of the DXC code generation and optimization pipeline shows
+that DXC allows Clang to handle resource arrays as if they were regular arrays
+of objects until after the LLVM optimization phase. Once all functions have been
+inlined in the generated code, DXC locates all references to resource arrays and
+replaces them with calls to initialize the resource handle.
+
+This process quickly becomes complex because local arrays are moved using
+`memcpy`, and since they are editable, correctly resolving all resource array
+accesses is not always straightforward. Given this, the incorrect behavior in
+the above example is not surprising.
+
+#### Resource Arrays in User-defined Structs
+
+Resources and resource arrays can also be included in user-defined structs which
+can be declared at local or global scope. While this feature will covered by a
+separate proposal (issues [#175](https://github.com/llvm/wg-hlsl/issues/175) and
+[#212](https://github.com/llvm/wg-hlsl/issues/212)), one example is included
+here for completeness:
+
+```
+struct S {
+  int a;
+  RWBuffer<float> P[5];
+};
+
+S s : register(u10);
+
+cbuffer CB {
+    S s_array[4];
+};
+
+RWBuffer<float> Out;
+
+[numthreads(4,1,1)]
+void main() {
+  Out[0] = s.P[2][0] + s_array[1].P[4][0];
+}
+```
+https://godbolt.org/z/Y8h3cWMov
+
+In DXC, when a user-defined struct contains an array of resources, each element
+of the array is treated as individual standalone resource instance. For `s.P[2]`
+in the example above the compiler creates a resource named `s.P.2` mapped to
+register `u12` with range `1`. Its handle initialization does not reflect that
```

Comment:
Does this affect the copy semantics if this resource array was passed into a function?

Reply:
As far as I know DXC is currently not doing any local array copies for function arguments (see Tex's comment https://github.com/llvm/wg-hlsl/pull/293#discussion_r2162743166).